torchvision/datasets/flowers102.py

```diff
@@ -0,0 +1,115 @@
+        transform (callable, optional): A function/transform that takes in an PIL image and returns a
+            transformed version. E.g, ``transforms.RandomCrop``.
+        target_transform (callable, optional): A function/transform that takes in the target and transforms it.
+    """
+
+    download_url_prefix = "https://www.robots.ox.ac.uk/~vgg/data/flowers/102/"
+    file_dict = {
+        "image": ("102flowers.tgz", "52808999861908f626f3c1f4e79d11fa"),
+        "label": ("imagelabels.mat", "e0620be6f572b9609742df49c70aed4d"),
+        "setid": ("setid.mat", "a5357ecc9cb78c4bef273ce3793fc85c"),
+    }
+    splits_map = {"train": "trnid", "valid": "valid", "test": "tstid"}
+
+    def __init__(
+        self,
+        root: str,
+        split: str = "train",
+        download: bool = True,
+        transform: Optional[Callable] = None,
+        target_transform: Optional[Callable] = None,
+    ) -> None:
+        super().__init__(root, transform=transform, target_transform=target_transform)
+        self._split = verify_str_arg(split, "split", ("train", "valid", "test"))
+        self._base_folder = Path(self.root) / "flowers-102"
+        self._images_folder = self._base_folder / "jpg"
+
+        if download:
+            self.download()
+
+        if not self._check_integrity():
+            raise RuntimeError("Dataset not found or corrupted. You can use download=True to download it")
+
+        from scipy.io import loadmat
+
+        set_ids = loadmat(self._base_folder / self.file_dict["setid"][0], squeeze_me=True)
+        image_ids = set_ids[self.splits_map[self._split]].tolist()
+
+        labels = loadmat(self._base_folder / self.file_dict["label"][0], squeeze_me=True)
+        image_id_to_label = dict(enumerate(labels["labels"].tolist(), 1))
+
+        self._labels = []
+        self._image_files = []
+        for image_id in image_ids:
+            self._labels.append(image_id_to_label[image_id])
+            self._image_files.append(self._images_folder / f"image_{image_id:05d}.jpg")
+
+    def __len__(self) -> int:
+        return len(self._image_files)
+
+    def __getitem__(self, idx) -> Tuple[Any, Any]:
+        image_file, label = self._image_files[idx], self._labels[idx]
+        image = PIL.Image.open(image_file).convert("RGB")
+
+        if self.transform:
+            image = self.transform(image)
+
+        if self.target_transform:
+            label = self.target_transform(label)
+
+        return image, label
+
+    def extra_repr(self) -> str:
+        return f"split={self._split}"
+
+    def _check_exists(self) -> bool:
+        return self._images_folder.exists() and self._images_folder.is_dir()
+
+    def _check_integrity(self):
+        for id in ["label", "setid"]:
+            filename, md5 = self.file_dict[id]
+            if not check_integrity(str(self._base_folder / filename), md5):
+                return False
+        return True
+
+    def download(self):
+        if not self._check_exists():
```

Comment:
Hi @pmeier , I move the folder existence check here. Otherwise if the `labels` or `setids` is missing, we will download the image archive.

Reply:
Could we perhaps simplify the logic here, and just download all the files if `_check_exists()` returns False?

I'm not sure that allowing the image archive to be there without the labels or setids is really useful?

Reply:
IIRC the discrepancy in `download` was introduced after my review. I agree with Nicolas here. We should have one method that checks the existence / integrity of all files and one download method that downloads everything if the former method returns `False.`

Reply:
Hi @pmeier and @NicolasHug , I've reverted the change back.